File: typescript/packages/management/src/tools/create_trigger.ts
import type { CreateTriggerRequest } from '@copass/core';

import type { ToolContext, ToolHandler } from '../registrar.js';

export const createTrigger: ToolHandler = async (
  ctx: ToolContext,
  input: Record<string, unknown>,
) => {
  const agentSlug = String(input.agent_slug);
  const dataSourceId = String(input.data_source_id);
  const eventType =
    typeof input.event_type_filter === 'string' && input.event_type_filter.length > 0
      ? input.event_type_filter
      : '*';
  const request: CreateTriggerRequest = {
    data_source_id: dataSourceId,
    event_type_filter: eventType,
  };
  if (
    typeof input.filter_config === 'object' &&
    input.filter_config !== null
  ) {
    request.filter_config = input.filter_config as Record<string, unknown>;
  }
  if (typeof input.rate_limit_per_hour === 'number') {
    request.rate_limit_per_hour = input.rate_limit_per_hour;
  }
  const trigger = await ctx.client.agents.triggers.create(
    ctx.sandboxId,
    agentSlug,
    request,
  );
  return { trigger };
};

File: python/copass-management/src/copass_management/tools/create_trigger.py
from __future__ import annotations

from typing import Any, Dict, TYPE_CHECKING

if TYPE_CHECKING:
    from copass_management.registrar import ToolContext


async def create_trigger(ctx: "ToolContext", input: Dict[str, Any]) -> Any:
    agent_slug = str(input["agent_slug"])
    data_source_id = str(input["data_source_id"])
    event_type = str(input.get("event_type_filter") or "*")
    kwargs: Dict[str, Any] = {
        "data_source_id": data_source_id,
        "event_type_filter": event_type,
    }
    if isinstance(input.get("filter_config"), dict):
        kwargs["filter_config"] = dict(input["filter_config"])
    if input.get("rate_limit_per_hour") is not None:
        kwargs["rate_limit_per_hour"] = int(input["rate_limit_per_hour"])
    trigger = await ctx.client.agents.triggers.create(
        ctx.sandbox_id, agent_slug, **kwargs,
    )
    return {"trigger": trigger}
